pkg/autoupdate/tuf_autoupdate.go

```diff
@@ -0,0 +1,211 @@
+		checkInterval:   60 * time.Second,
+		errorCounter:    make([]int64, 0), // For now, the error counter is a simple list of timestamps when errors occurred
+		logger:          log.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(ta)
+	}
+
+	return ta, nil
+}
+
+func (ta *TufAutoupdater) Run(opts ...legacytuf.Option) (stop func(), err error) {
+	go ta.loop()
+
+	return ta.stop, nil
+}
+
+func (ta *TufAutoupdater) RollingErrorCount() int {
+	ta.lock.RLock()
+	defer ta.lock.RUnlock()
+
+	oneDayAgo := time.Now().Add(-24 * time.Hour).Unix()
+	errorCount := 0
+	for _, errorTimestamp := range ta.errorCounter {
+		if errorTimestamp < oneDayAgo {
+			continue
+		}
+		errorCount += 1
+	}
+
+	return errorCount
+}
+
+func (ta *TufAutoupdater) loop() error {
+	checkTicker := time.NewTicker(ta.checkInterval)
+
+	for {
+		select {
+		case <-checkTicker.C:
+			if err := ta.checkForUpdate(); err != nil {
+				ta.lock.Lock()
+				ta.errorCounter = append(ta.errorCounter, time.Now().Unix())
+				ta.lock.Unlock()
+
+				level.Debug(ta.logger).Log("msg", "error checking for update", "err", err)
+			}
+		case <-ta.interrupt:
+			level.Debug(ta.logger).Log("msg", "received interrupt, stopping")
+			return nil
+		}
+	}
+}
+
+func (ta *TufAutoupdater) stop() {
+	ta.interrupt <- struct{}{}
+}
+
+func (ta *TufAutoupdater) checkForUpdate() error {
+	_, err := ta.metadataClient.Update()
+	if err != nil {
+		return fmt.Errorf("could not update metadata: %w", err)
+	}
+
+	// Find the newest release for our channel -- right now for logging purposes only
+	targets, err := ta.metadataClient.Targets()
+	if err != nil {
+		return fmt.Errorf("could not get complete list of targets: %w", err)
+	}
+
+	targetReleaseFile := fmt.Sprintf("%s/%s/%s/release.json", strings.TrimSuffix(ta.binary, ".exe"), ta.operatingSystem, ta.channel)
+	for targetName, target := range targets {
+		if targetName != targetReleaseFile {
+			continue
+		}
+
+		// We found the release file that matches our OS and binary. Evaluate it
+		// to see if we're on this latest version.
+		type releaseFileCustomMetadata struct {
+			Target string `json:"target"`
+		}
+
+		var custom releaseFileCustomMetadata
+		if err := json.Unmarshal(*target.Custom, &custom); err != nil {
+			return fmt.Errorf("could not unmarshal release file custom metadata: %w", err)
+		}
+
+		level.Debug(ta.logger).Log(
+			"msg", "checked most up-to-date release from TUF",
+			"launcher_version", version.Version().Version,
+			"release_version", ta.versionFromTarget(custom.Target),
+			"channel", ta.channel,
+		)
+
+		break
+	}
+
+	return nil
+}
+
+func (ta *TufAutoupdater) versionFromTarget(target string) string {
+	strippedBinary := strings.TrimSuffix(ta.binary, ".exe")
+
+	// The target is in the form `launcher/linux/launcher-0.13.6.tar.gz` -- trim the prefix and the file extension to return the version
+	prefixToTrim := fmt.Sprintf("%s/%s/%s-", strippedBinary, ta.operatingSystem, strippedBinary)
```

Comment:
Urk. I've lost site of how we track archtecture

Reply:
I don't believe it's present in the TUF/dl.kolide.co stuff

Reply:
I also think this is something we'll need. There have been a few times that folks have asked for arm64 on linux / windows. I think number will continue to grow. @RebeccaMahany , do you think this would be difficult to bolt on later?